ledger/src/shred.rs

```diff
@@ -510,12 +518,17 @@ impl Shred {
     }
 
     // Returns true if the shred passes sanity checks.
-    pub(crate) fn sanitize(&self) -> bool {
-        self.erasure_block_index().is_some()
+    // TODO: Should return why sanitize failed!
+    pub fn sanitize(&self) -> bool {
+        self.payload.len() <= SHRED_PAYLOAD_SIZE
+            && self.erasure_block_index().is_some()
             && match self.shred_type() {
                 ShredType::Data => {
-                    self.parent().is_ok()
-                        && usize::from(self.data_header.size) <= self.payload.len()
+                    let size = usize::from(self.data_header.size);
+                    self.index() < MAX_DATA_SHREDS_PER_SLOT as u32
+                        && self.parent().is_ok()
+                        && size <= self.payload.len()
+                        && DATA_SHRED_SIZE_RANGE.contains(&size)
```

Comment:
Could we add a comment explaining this? I'm a bit confused. 

`DATA_SHRED_SIZE_RANGE` is defined as `SHRED_DATA_OFFSET..=SHRED_DATA_OFFSET + SIZE_OF_DATA_SHRED_PAYLOAD;`, so it's the offset at which the headers end and the data payload begins. 

Why must the `self.data_header.size` be in this range? Shouldn't it be in the range `0..SIZE_OF_DATA_SHRED_PAYLOAD`?

Reply:
Because `self.data_header.size` includes the size of headers as well:
https://github.com/solana-labs/solana/blob/38bdb401a/ledger/src/shred.rs#L291-L293

I have included comments for both `DATA_SHRED_SIZE_RANGE` and `DataShredHeader.size`.

Reply:
ah, got it, thanks!